presto-main/src/main/java/com/facebook/presto/execution/TaskThresholdMemoryRevokingScheduler.java

```diff
@@ -0,0 +1,147 @@
+import com.facebook.presto.operator.OperatorContext;
+import com.facebook.presto.sql.analyzer.FeaturesConfig;
+import com.google.common.annotations.VisibleForTesting;
+
+import javax.annotation.Nullable;
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+import javax.inject.Inject;
+
+import java.util.Collection;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Supplier;
+
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+public class TaskThresholdMemoryRevokingScheduler
+{
+    private static final Logger log = Logger.get(MemoryRevokingScheduler.class);
+    private final Supplier<? extends Collection<SqlTask>> currentTasksSupplier;
+    private final ScheduledExecutorService taskManagementExecutor;
+    private final long maxRevocableMemoryPerTask;
+
+    @Nullable
+    private ScheduledFuture<?> scheduledFuture;
+
+    private final AtomicBoolean checkPending = new AtomicBoolean();
+
+    @Inject
+    public TaskThresholdMemoryRevokingScheduler(
+            SqlTaskManager sqlTaskManager,
+            TaskManagementExecutor taskManagementExecutor,
+            FeaturesConfig config)
+    {
+        this(
+                requireNonNull(sqlTaskManager, "sqlTaskManager cannot be null")::getAllTasks,
+                requireNonNull(taskManagementExecutor, "taskManagementExecutor cannot be null").getExecutor(),
+                config.getMaxRevocableMemoryPerTask());
+        log.debug("Using TaskThresholdMemoryRevokingScheduler spilling strategy");
+    }
+
+    @VisibleForTesting
+    TaskThresholdMemoryRevokingScheduler(
+            Supplier<? extends Collection<SqlTask>> currentTasksSupplier,
+            ScheduledExecutorService taskManagementExecutor,
+            long maxRevocableMemoryPerTask)
+    {
+        this.currentTasksSupplier = requireNonNull(currentTasksSupplier, "currentTasksSupplier is null");
+        this.taskManagementExecutor = requireNonNull(taskManagementExecutor, "taskManagementExecutor is null");
+        this.maxRevocableMemoryPerTask = maxRevocableMemoryPerTask;
+    }
+
+    @PostConstruct
+    public void start()
+    {
+        registerTaskMemoryPeriodicCheck();
+    }
+
+    private void registerTaskMemoryPeriodicCheck()
+    {
+        this.scheduledFuture = taskManagementExecutor.scheduleWithFixedDelay(() -> {
+            try {
+                revokeHighMemoryTasksIfNeeded();
+            }
+            catch (Throwable e) {
+                log.error(e, "Error requesting task memory revoking");
+            }
+        }, 1, 1, SECONDS);
+    }
+
+    @PreDestroy
+    public void stop()
+    {
+        if (scheduledFuture != null) {
+            scheduledFuture.cancel(true);
+            scheduledFuture = null;
+        }
```

Comment:
This is not thread-safe

Reply:
Since this class is injected as a SINGLETON + this part is only called on `@PreDestroy`, this method won't be called until this class is destroyed (We never call `stop()` manually). This won't happen until Presto shuts down? I can't imagine any other case in which we will destroy these schedulers. This method only exists to stop leaking this future for tests I'm guessing (since a similar method is also present in `MemoryRevokingScheduler`). Let me know if knowing this, we still need thread safety for this variable.